client/asset/eth/rpcclient.go

```diff
@@ -256,6 +261,64 @@ func (c *rpcclient) initiate(txOpts *bind.TransactOpts, netID int64, refundTimes
 	return c.es.Initiate(txOpts, big.NewInt(refundTimestamp), secretHash, *participant)
 }
 
+// initiateGas checks the amount of gas that is used for a call to the initiate function.
+func (c *rpcclient) initiateGas(ctx context.Context, contractAddress *common.Address) (uint64, error) {
+	parsedAbi, err := abi.JSON(strings.NewReader(swap.ETHSwapABI))
+	if err != nil {
+		return 0, err
+	}
+	refundTimestamp := time.Now().Unix()
+	var secret [32]byte
+	copy(secret[:], encode.RandomBytes(32))
+	secretHash := sha256.Sum256(secret[:])
+	// just an arbitrary address
+	participant := common.HexToAddress("345853e21b1d475582E71cC269124eD5e2dD3422")
+	data, err := parsedAbi.Pack("initiate", big.NewInt(refundTimestamp), secretHash, participant)
+	if err != nil {
+		return 0, nil
+	}
+
+	msg := ethereum.CallMsg{
+		From:  participant,
+		To:    contractAddress,
+		Gas:   0,
+		Value: big.NewInt(100000000),
+		Data:  data,
+	}
+	gas, err := c.ec.EstimateGas(ctx, msg)
+	if err != nil {
+		return 0, err
+	}
+
+	return gas, nil
```

Comment:
I would prefer if the rpcclient methods had as little logic as possible. So, the caller do all the set up and the rpcclient just calls this last bit. This is because we can test better, as we cannot make a fake *rpcclient struct, but we can make a fake interface that implements the node interface and test the code above this block there. Does that make sense? The client can just accept the msg as an arg. If this sounds silly let me know, it's not set in stone.

Reply:
I don't think having this logic here really restricts testing. If we end up using these two functions outside of the test code, we could easily just return predetermined values for how much gas is used by the two functions.

I was thinking that this functionality that is missing from the autogenerated ETHSwap code. It would be nice if the go code generated from the solidity allowed you not only to call the functions, but also to check how much gas they would use before you call them. SInce `*swap.ETHSwap` is part of the `rpcclient` struct, I think this is a good place to add that extended functionality.

Reply:
If you want to add it in this particular file, you could also make it a separate function... so it's testable. You don't use `c` until the very end at `c.ec.EstimateGas`